lib/estimator/estimator.rb

```diff
@@ -0,0 +1,89 @@
+      ship_params[:packages].each do |package|
+        package_info = pack_items(package)
+        active_origin = ActiveShipping::Location.new(package[:origin])
+        active_destination = destination(ship_params)
+        active_package = ActiveShipping::Package.new(package_info[:weight], package_info[:dimensions], :units => :imperial, :value => 10) #had to add value for delivery date estimate
+
+        package_hash = {origin: active_origin,
+                        destination: active_destination,
+                        package: active_package
+                        }
+        shipment_array << package_hash
+      end
+      return shipment_array
+    end
+
+    def self.pack_items(package)
+      #different items from the same merchant will be packed together into one package
+      #a simple way to do this is to find the longest length and and add together the widths and heights. return the total weight and the dimensions of this package.
+      #this method assumes the length is the longest dimension
+      weight = package[:package_items].map {|item| item[:weight].to_i}.sum
+      longest_package = package[:package_items].max_by{|item| item[:length]}
+      length = longest_package[:length].to_i
+      width = package[:package_items].map {|item| item[:width].to_i}.sum
+      height =  package[:package_items].map {|item| item[:height].to_i}.sum
+      package_info = {weight: weight, dimensions: [length, width, height]}
+      return package_info
+    end
+
+    def self.destination(ship_params)
+      # keeping destination as a separate method since that's how params will come in from wetsy.
+      #would have been better maybe to keep it more general for our api, so that packages can have any destination. But that's ok.
+      ActiveShipping::Location.new(ship_params[:destination])
+    end
+
+    def self.get_rates(shipment_array, carrier)
+      #rates is an array of hashes with different shipping services and their cost.
+      #total_carrier_rates creates as array summing the cost of each of the packages for the same service
+      #this is where are are taking all the packages from each different merchant and summing their shipping costs
+      #error handling takes care of things when the zip code is invalid, etc
+      carrier_rates = []
+        shipment_array.each do |shipment|
+            response = carrier.find_rates(shipment[:origin], shipment[:destination], shipment[:package])
+            sorted_rates = (response.rates.sort_by(&:price).collect {|rate| [rate.service_name, rate.price]}).to_h
+            carrier_rates << sorted_rates
+        end
+    rescue ActiveShipping::ResponseError
+      @api_call_ok = false
+    else
+      total_carrier_rates = Hash.new(0)
+      carrier_rates.each { |subhash| subhash.each { |service, cost| total_carrier_rates[service] += cost } }
+      return total_carrier_rates
+    end
+
+    def self.assemble_quote(ups, usps)
+      {"UPS" => ups, "USPS" => usps}
+    end
+  end
+end
```

Comment:
This class is doing A LOT. I would suggest separating the functionality out into multiple classes, specifically one class to act as a client file (to be solely responsible for connecting to the Active Shipping API) and a separate class to transform the data you get back from the Active Shipping API to a json format that your service will return to the client. This helps organize your thought process and makes it much easier to add on functionality when you get to that point in a project.